providers/file/provider.go

```diff
@@ -0,0 +1,292 @@
+		p.opts.KubeconfigFiles, p.opts.KubeconfigDirs = p.defaultKubeconfigPaths()
+	}
+
+	if len(p.opts.KubeconfigGlobs) == 0 {
+		p.opts.KubeconfigGlobs = DefaultKubeconfigGlobs
+	}
+
+	p.log = log.Log.WithName("file-cluster-provider")
+	p.clusters = make(map[string]cluster.Cluster)
+	p.clusterCancel = make(map[string]func())
+
+	fmt.Printf("file cluster provider initialized %q %q\n", p.opts.KubeconfigFiles, p.opts.KubeconfigDirs)
+
+	return p, nil
+}
+
+// Provider is a multicluster.Provider that loads clusters from
+// kubeconfig files or directories on disk.
+type Provider struct {
+	opts Options
+
+	log logr.Logger
+
+	clustersLock  sync.RWMutex
+	clusters      map[string]cluster.Cluster
+	clusterCancel map[string]func()
+}
+
+// Run starts the provider and updates the clusters and is blocking.
+func (p *Provider) Run(ctx context.Context, mgr mcmanager.Manager) error {
+	if err := p.run(ctx, mgr); err != nil {
+		return fmt.Errorf("initial update failed: %w", err)
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to create file watcher: %w", err)
+	}
+	defer watcher.Close()
+
+	for _, file := range p.opts.KubeconfigFiles {
+		// Watching the directory instead of kubeconfig files as
+		// watching non-existing files is not supported in fsnotify.
+		if err := watcher.Add(filepath.Dir(file)); err != nil {
+			return fmt.Errorf("failed to watch parent dir of kubeconfig file %q: %w", file, err)
+		}
+	}
+
+	for _, dir := range p.opts.KubeconfigDirs {
+		if err := watcher.Add(dir); err != nil {
+			return fmt.Errorf("failed to watch kubeconfig directory %q: %w", dir, err)
+		}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case event, ok := <-watcher.Events:
+			if !ok {
+				return fmt.Errorf("file watcher closed")
+			}
+			p.log.Info("received fsnotify event", "event", event)
+			// Only updating from a single file would be possible but
+			// would also require to track which cluster belongs to
+			// which file.
+			// Instead clusters are just updated from all files.
+			if err := p.run(ctx, mgr); err != nil {
+				p.log.Error(err, "failed to update clusters after file change")
+			}
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				return fmt.Errorf("file watcher errors channel closed")
+			}
+			p.log.Error(err, "file watcher error")
+		}
+	}
+}
+
+// RunOnce performs a single update of the clusters.
+func (p *Provider) RunOnce(ctx context.Context, mgr mcmanager.Manager) error {
+	return p.run(ctx, mgr)
+}
+
+func (p *Provider) addCluster(ctx context.Context, mgr mcmanager.Manager, name string, cl cluster.Cluster) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	p.clustersLock.Lock()
+	p.clusters[name] = cl
+	p.clusterCancel[name] = cancel
+	p.clustersLock.Unlock()
+
+	go func() {
+		if err := cl.Start(ctx); err != nil {
+			p.log.Error(err, "error in cluster", "name", name)
+		}
+		p.removeCluster(name)
```

Comment:
Wouldn't this goroutine attempt to remove the cluster when we are canceling the context (i.e. shutting down the controller-manager)? Is that something to care about?